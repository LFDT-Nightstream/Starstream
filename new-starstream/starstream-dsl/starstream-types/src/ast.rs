//! Abstract syntax tree for the pared-down Starstream DSL.

use chumsky::span::SimpleSpan;
use serde::Serialize;

/// Entire program: a sequence of statements.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Program {
    pub statements: Vec<Statement>,
}

/// Statements allowed by the current grammar.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Statement {
    VariableDeclaration {
        name: Identifier,
        value: Spanned<Expr>,
    },
    Assignment {
        target: Identifier,
        value: Spanned<Expr>,
    },
    If {
        condition: Spanned<Expr>,
        then_branch: Block,
        else_branch: Option<Block>,
    },
    While {
        condition: Spanned<Expr>,
        body: Block,
    },
    Block(Block),
    Expression(Spanned<Expr>),
}

/// `{ statement* }`
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Block {
    pub statements: Vec<Statement>,
}

/// Expressions with precedence encoded via parser, not the AST shape.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Expr {
    Literal(Literal),
    Identifier(Identifier),
    Unary {
        op: UnaryOp,
        expr: Box<Spanned<Expr>>,
    },
    Binary {
        op: BinaryOp,
        left: Box<Spanned<Expr>>,
        right: Box<Spanned<Expr>>,
    },
    Grouping(Box<Spanned<Expr>>),
}

#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Literal {
    Integer(i64),
    Boolean(bool),
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
pub enum UnaryOp {
    Negate,
    Not,
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
pub enum BinaryOp {
    Multiply,
    Divide,
    Remainder,
    Add,
    Subtract,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    And,
    Or,
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq, Hash)]
pub struct Identifier {
    pub name: String,
    #[serde(skip)]
    pub span: Option<SimpleSpan>,
}

impl Identifier {
    pub fn new(name: impl Into<String>, span: Option<SimpleSpan>) -> Self {
        Self {
            name: name.into(),
            span,
        }
    }
}

/// Helper for attaching source spans to AST nodes.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Spanned<T> {
    pub node: T,
    #[serde(skip)]
    pub span: SimpleSpan,
}

impl<T> Spanned<T> {
    pub fn new(node: T, span: SimpleSpan) -> Self {
        Self { node, span }
    }

    pub fn none(node: T) -> Spanned<T> {
        Spanned {
            node,
            span: SimpleSpan {
                start: 0,
                end: 0,
                context: (),
            },
        }
    }
}
