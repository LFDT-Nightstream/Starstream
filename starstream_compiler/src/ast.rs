//! AST types describing a Starstream source file.

#[derive(Clone, Debug, Default)]
pub struct StarstreamProgram {
    pub items: Vec<ProgramItem>,
}

#[derive(Clone, Debug)]
pub enum ProgramItem {
    Utxo(Utxo),
    Script(Script),
    Token(Token),
}

#[derive(Clone, Debug)]
pub struct Utxo {
    pub name: Identifier,
    pub items: Vec<UtxoItem>,
}

#[derive(Clone, Debug)]
pub enum UtxoItem {
    Abi(Abi),
    Main(Main),
    Impl(Impl),
    Storage(Storage),
}

#[derive(Clone, Debug)]
pub struct Main {
    pub type_sig: Option<OptionallyTypedBindings>,
    pub block: Block,
}

#[derive(Clone, Debug)]
pub struct Token {
    pub name: Identifier,
    pub items: Vec<TokenItem>,
}

#[derive(Clone, Debug)]
pub enum TokenItem {
    Bind(Bind),
    Unbind(Unbind),
    Abi(Abi),
    Mint(Mint),
}

#[derive(Clone, Debug)]
pub struct Bind(pub Block);

#[derive(Clone, Debug)]
pub struct Unbind(pub Block);

#[derive(Clone, Debug)]
pub struct Mint(pub Block);

#[derive(Clone, Debug)]
pub struct Impl {
    pub name: Identifier,
    pub definitions: Vec<FnDef>,
}

#[derive(Clone, Debug)]
pub struct Script {
    pub definitions: Vec<FnDef>,
}

#[derive(Clone, Debug)]
pub struct Storage {
    pub bindings: TypedBindings,
}

#[derive(Clone, Debug)]
pub struct Sig {
    pub name: Identifier,
    pub input_types: Vec<Type>,
    pub output_type: Option<Type>,
}

#[derive(Clone, Debug)]
pub struct FnSig(pub Sig);

#[derive(Clone, Debug)]
pub struct FnDef {
    pub name: Identifier,
    pub inputs: OptionallyTypedBindings,
    pub output: Option<Type>,
    pub body: Block,
}

#[derive(Clone, Debug)]
pub enum EffectSig {
    EffectSig(Sig),
    EventSig(Sig),
    ErrorSig(Sig),
}

#[derive(Clone, Debug)]
pub enum AbiElem {
    FnSig(FnSig),
    EffectSig(EffectSig),
}

#[derive(Clone, Debug)]
pub struct Abi {
    pub values: Vec<AbiElem>,
}

#[derive(Clone, Debug)]
pub struct Identifier(pub String);

#[derive(Clone, Debug)]
pub enum Type {
    BaseType(Identifier, Option<Vec<Type>>),
    Object(TypedBindings),
    FnType(TypedBindings, Option<Box<Type>>),
}

#[derive(Clone, Debug)]
pub enum Statement {
    BindVar {
        var: Identifier,
        value: Expr,
        mutable: bool,
    },
    Return(Option<Expr>),
    Resume(Option<Expr>),
    Assign(Identifier, Expr),
    With(Block, Vec<(Effect, Block)>),
    While(Expr, LoopBody),
    Loop(LoopBody),
}

#[derive(Clone, Debug)]
pub enum LoopBody {
    Statement(Box<Statement>),
    Block(Block),
    Expr(Expr),
}

#[derive(Clone, Debug)]
pub enum Expr {
    Equals(Box<Self>, Box<Self>),
    NotEquals(Box<Self>, Box<Self>),
    LessThan(Box<Self>, Box<Self>),
    Add(Box<Self>, Box<Self>),
    Or(Box<Self>, Box<Self>),
    Sub(Box<Self>, Box<Self>),
    Pow(Box<Self>, Box<Self>),
    Mul(Box<Self>, Box<Self>),
    And(Box<Self>, Box<Self>),
    Div(Box<Self>, Box<Self>),
    Neg(Box<Self>),
    PrimaryExpr(
        PrimaryExpr,
        Option<Arguments>,
        Vec<(Identifier, Option<Arguments>)>,
    ),
    BlockExpr(BlockExpr),
}

#[derive(Clone, Debug)]
pub enum BlockExpr {
    IfThenElse(Box<Expr>, Box<Block>, Option<Box<Block>>),
    Block(Block),
}

#[derive(Clone, Debug)]
pub enum PrimaryExpr {
    Null,
    Number(f64),
    Bool(bool),
    Ident(Identifier),
    ParExpr(Box<Expr>),
    Yield(Box<Expr>),
    Raise(Box<Expr>),
    Object(Type, Vec<(Identifier, Expr)>),
}

#[derive(Clone, Debug)]
pub enum ExprOrStatement {
    Expr(Expr),
    Statement(Statement),
}

#[derive(Clone, Debug)]
pub enum Block {
    Chain {
        head: Box<ExprOrStatement>,
        tail: Box<Block>,
    },
    Close {
        semicolon: bool,
    },
}

#[derive(Clone, Debug)]
pub struct Arguments {
    pub xs: Vec<Expr>,
}

#[derive(Clone, Debug)]
pub struct OptionallyTypedBindings {
    pub values: Vec<(Identifier, Option<Type>)>,
}

#[derive(Clone, Debug)]
pub struct TypedBindings {
    pub values: Vec<(Identifier, Type)>,
}

#[derive(Clone, Debug)]
pub struct Effect {
    pub ident: Identifier,
    pub type_sig: OptionallyTypedBindings,
}
