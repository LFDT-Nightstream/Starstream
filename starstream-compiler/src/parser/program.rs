use chumsky::prelude::*;
use starstream_types::ast::Program;

use super::{context::Extra, statement};

/// Get a parser for a Starstream source file.
pub fn parser<'a>() -> impl Parser<'a, &'a str, Program, Extra<'a>> {
    statement::parser()
        .padded()
        .repeated()
        .collect::<Vec<_>>()
        .then_ignore(end())
        .map(|statements| Program { statements })
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;

    macro_rules! assert_program_snapshot {
        ($code:expr) => {{
            let program = parser()
                .parse(indoc! { $code })
                .into_result()
                .expect("program should parse");

            insta::with_settings!({
                description => format!("Code:\n\n{}", indoc! { $code }),
                omit_expression => true,
                prepend_module_to_snapshot => true,
            }, {
                insta::assert_debug_snapshot!(program);
            });
        }};
    }

    #[test]
    fn sequential_statements() {
        assert_program_snapshot!(
            r#"
            let count = 0;
            count = count + 1;
            false;
            "#
        );
    }

    #[test]
    fn conditional_program() {
        assert_program_snapshot!(
            r#"
            let flag = true;
            if (flag) { let answer = 42; } else { answer = 0; }
            "#
        );
    }
}
