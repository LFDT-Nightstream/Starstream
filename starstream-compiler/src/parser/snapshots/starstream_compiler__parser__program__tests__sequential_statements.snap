---
source: starstream-compiler/src/parser/program.rs
description: "Code:\n\nlet count = 0;\ncount = count + 1;\nfalse;\n"
---
Program {
    statements: [
        VariableDeclaration {
            name: Identifier {
                name: "count",
                span: Some(
                    4..9,
                ),
            },
            value: Spanned {
                node: Literal(
                    Integer(
                        0,
                    ),
                ),
                span: 12..13,
            },
        },
        Assignment {
            target: Identifier {
                name: "count",
                span: Some(
                    15..20,
                ),
            },
            value: Spanned {
                node: Binary {
                    op: Add,
                    left: Spanned {
                        node: Identifier(
                            Identifier {
                                name: "count",
                                span: Some(
                                    23..28,
                                ),
                            },
                        ),
                        span: 23..29,
                    },
                    right: Spanned {
                        node: Literal(
                            Integer(
                                1,
                            ),
                        ),
                        span: 31..32,
                    },
                },
                span: 23..32,
            },
        },
        Expression(
            Spanned {
                node: Literal(
                    Boolean(
                        false,
                    ),
                ),
                span: 34..39,
            },
        ),
    ],
}
